Remove using namespace libMesh

GiudGiud opened the pull request:
closes #28499

patches:
- [ ] SAM https://git-nse.egs.anl.gov/SAM/SAM/-/merge_requests/971
- [x] falcon https://github.com/idaholab/falcon/pull/148
- [x] pronghorn https://github.inl.gov/ncrc/pronghorn/pull/1276
- [x] griffin https://github.inl.gov/ncrc/griffin/pull/2147

I used the techniques below to make this a little painful. Some of these are not "best practices" so I can avoid these shortcuts, but at the cost of a lot more work most likely.

For source files:
- added `libMesh::` in a lot of source files initially. This was taking forever so I...
- added `using namespace libMesh` at the top of source files. This is probably the cleaner way to do this. Unfortunately this quick and easy solution should not be done in headers. And because of unity build I probably have to do this again a few times to get every single source file that actually needs it

For headers:
- I did add `libMesh::` in a lot of headers as well. This was also taking forever, it would be needed in thousands of places
- so I started adding `using libMesh::Real/ .../` in common headers. Namely MooseTypes.h, MooseMesh.h, AuxiliarySystem.h .... This is not a recommended practice but I don't really see us having `libMesh::Real` (or `using libMesh::Real;`) in every header?? But maybe we should. Since too many files were needed, I ended up centralizing in libMeshReduceNamespace.h
- for the nested namespaces, I added something like `namespace Threads = libMesh::Threads` which effectively nullifies the effect of removing `using namespace libMesh`. So we keep the same degree of contamination with regards to these nested namespaces of libMesh (e.g. you can't use a Threads namespace with overlapping symbols from another library)
- then to pass CI of the other apps I basically had to add a couple other `using libMesh::` at the end of the libMeshReducedNamespace.h. The alternative is to patch every single app, which is not very reasonable at this point. Unfortunately most apps do not have a `XYZnameoftheapp.h` header they include everywhere.

In any case, even if we don't like any of these practices, we can fix these gradually, unlike the removal of `using namespace libMesh` which can only be brutal

I think I can improve what I did in `libMeshReducedNamespace.h` by using forward declarations. I ll try that
EDIT: done

@grmnptr

GiudGiud committed:
Remove libMesh namespace from framework: mostly headers
- declare sub-namespaces of libMesh as their own namespaces
- add libMesh:: in front of many objects
- add using libMesh::many_common_ones
refs #28499

GiudGiud committed:
Remove libMesh namespace from framework source:
- heavily rely on using namespace libMesh
- for some files, add libMesh:: everywhere
refs #28499

GiudGiud committed:
Adapt framework test objects to libMesh namespace removal
refs #28499

GiudGiud committed:
Adapt modules to removal of libMesh namespace
rfs #28499

GiudGiud commented:
@dschwen @lindsayad

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2494113/), step Clang format on 8018a7f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0a0f83809db247cb66248567225b398ffabca9ba`

GiudGiud committed:
Adapt MOOSE unit tests to removal of 'using namespace libMesh' in global scope

GiudGiud committed:
Misc change to pass precheck

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2495986/), step Clang format on 9ea7ecb wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a35543bb275d046234e6544b1e4c11563b57c710`

GiudGiud committed:
Add more of libMesh namespace to pass downstream app tests

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2496457/), step Clang format on a4a4951 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a35543bb275d046234e6544b1e4c11563b57c710`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2497863/), step Clang format on 18448c7 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b3924ece4ed556a7ca9811a6e046eb5c4503df72`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2498844/), step Clang format on b2c1d40 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b3924ece4ed556a7ca9811a6e046eb5c4503df72`

GiudGiud commented on framework/include/base/Moose.h:
```diff
@@ -9,9 +9,11 @@
 #include "libmesh/perf_log.h"
 #include "libmesh/libmesh_common.h"
 #include "XTermConstants.h"
+#include "libmesh/utility.h" // to declare namespace
```
self-note to remove this

roystgnr replied:
You want it just for the namespace declaration?  You can just put `namespace libMesh {}` in MOOSE code to get that.

GiudGiud replied:
yup! I ended up figuring this out later on

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2498990/), step Docs: sync website on cc51284 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28842/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2499005/), step Generate coverage on cc51284 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.05%</td>
      <td>85.04%</td>
      <td>-0.00%</td>
      <td>70.10%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>106298</td>
      <td>106311</td>
      <td>+13</td>
      <td>340</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18691</td>
      <td>18696</td>
      <td>+5</td>
      <td>145</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/framework)


## Modules coverage

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.71%</td>
      <td>88.71%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4351</td>
      <td>4352</td>
      <td>+1</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>554</td>
      <td>554</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/heat_transfer)

### Level set

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.45%</td>
      <td>85.49%</td>
      <td>+0.04%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>323</td>
      <td>324</td>
      <td>+1</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>55</td>
      <td>55</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/level_set)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/level_set)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.48%</td>
      <td>84.48%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>17002</td>
      <td>17002</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3123</td>
      <td>3123</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/navier_stokes)

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.01%</td>
      <td>86.01%</td>
      <td>+0.00%</td>
      <td>75.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13673</td>
      <td>13675</td>
      <td>+2</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2224</td>
      <td>2224</td>
      <td>-</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/phase_field)

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.30%</td>
      <td>95.30%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>11336</td>
      <td>11336</td>
      <td>-</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>559</td>
      <td>559</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/porous_flow)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/porous_flow)

### Ray tracing

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.33%</td>
      <td>95.33%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4364</td>
      <td>4364</td>
      <td>-</td>
      <td>7</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>214</td>
      <td>214</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/ray_tracing)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/ray_tracing)

### Reactor

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.65%</td>
      <td>93.65%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7613</td>
      <td>7613</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>516</td>
      <td>516</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/reactor)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/reactor)

### Richards

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.11%</td>
      <td>93.11%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3324</td>
      <td>3324</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>246</td>
      <td>246</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/richards)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/richards)

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.91%</td>
      <td>84.91%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>28158</td>
      <td>28158</td>
      <td>-</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>5006</td>
      <td>5006</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/solid_mechanics)

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.68%</td>
      <td>90.68%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>8093</td>
      <td>8093</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>832</td>
      <td>832</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/stochastic_tools)

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.35%</td>
      <td>88.35%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13910</td>
      <td>13911</td>
      <td>+1</td>
      <td>15</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1834</td>
      <td>1834</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/thermal_hydraulics)

### Xfem

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28842 cc5128</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.38%</td>
      <td>82.38%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6907</td>
      <td>6907</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1477</td>
      <td>1477</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage_diff/xfem)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28842/coverage/xfem)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28842/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 70.10% is less than the suggested 90.0%
- `phase_field` new line coverage rate 75.00% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
All (needed with the huge concessions on the namespace) patches are merged this should be good to go.

This will collect merge conflicts and unexpected failures (from new uses of objects that are not in the namespace) like no other PR so it'd be great to process this fast

loganharbour reviewed (dismissed):
At the very least needs a newsletter entry. this is pretty significant

Added more testing for nonstandard configurations. I suspect we might break a few apps with nonstandard configs (that aren't tested in on-moose app tests), but we can deal with that on app devels as they pop up

GiudGiud committed:
Add newsletter entry for libMesh namespace inclusion removal

lindsayad reviewed (commented):
seems like a great start

GiudGiud commented:
Thanks thanks
Looks like the jit stuff fails on Mac I ll try to figure it out

GiudGiud committed:
Fix AD FParser compilation error by including a lot less headers

GiudGiud committed:
Address failures from removing include mooseError.h from ADReal.h
Include it instead in MooseTypes.h which is including ADReal.h and was bringing in mooseError to everyone as it is commonly included

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2580924/), step Clang format on bea34b8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 404762b58321db088558adad176c13c466bd2585`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2581260/), step Clang format on 2a0a6c5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28842/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 404762b58321db088558adad176c13c466bd2585`

GiudGiud commented:
This should be working now. 
The issue was those the elem_range and nodal_range libmesh header causes a failure in the JITCompile. I am not sure why, but I worked around it by making sure they were not included. This required cutting the include chain
ADReal.h includes MooseError.h includes Moose.h include libMeshReducedNamespace.h includes the range headers.

I cut it at the first link.
Not including Moose.h in MooseError.h does not work (too many Moose:: namespace objects in mooseError)
Not including libMeshReducedNamespace.h in Moose.h does not work (reduced namespace is intended to facilitate the transition from having `using namespace libMesh` in Moose.h
Not including the range headers in the reduced namespace header would require a ton of work in both Moose and the downstream apps to prefix (Const|)Elem/NodeRange with libMesh::.

This last option is the intended path forward after this PR imo.

GiudGiud commented:
I have stashed the newsletter contribution. I ll re-add it once Casey's newsletter is merged because it would conflict

lindsayad commented on framework/include/base/libMeshReducedNamespace.h:
```diff
@@ -0,0 +1,418 @@
+#include "libmesh/libmesh.h"
+#include "libmesh/id_types.h"
+#include "libmesh/int_range.h"
+#include "libmesh/vector_value.h"
+#include "libmesh/tensor_value.h"
```
I don't understand some of these includes, e.g. these. A comment about why they are included would be helpful

GiudGiud replied:
this one brings a bunch of typedef so it must be why
```
typedef TensorValue<Real>   RealTensorValue;
typedef TensorValue<Number> NumberTensorValue;
typedef RealTensorValue     RealTensor;
typedef NumberTensorValue   Tensor;
```
there is no forward-declaration of typedefs. If I want to put RealTensor in the local namespace (using libMesh::RealTensor), I need to either add the typedef again or include the header.
including the header is better to reduce code duplication

GiudGiud replied:
I ll double check these headers and separate them in two lists:
- needed to get a function without re-declaring it
- needed to get a typedef without re-defining it

roystgnr replied:
Relying on a forward declaration instead of a header from a library can be dangerous, since it turns some internal library refactoring (e.g. adding a new optional template argument or changing a typedef) into API-breaking refactoring that would otherwise have been fine.  I think it falls somewhere between "heavily discouraged" and "non-standards-conforming" for `std::` classes for this reason, with the standard going so far as to create `<iosfwd>` for the one case where people were *really* ticked about the long compile times of the full header.

For libMesh usage in MOOSE, on the other hand, I'd say it's fine to do forward declarations.  If we ever break compatibility with them then we'll fix it at submodule upgrade time.

On the gripping hand*, the only real benefit of forward declarations over header inclusion is that they reduce compile time, and I don't think the reduction is a large one for unity builds where every header is likely to end up included from another .C file anyway.  So either way is fine for us.

\* did I just use that idiom twice in one day? feels like I'm catching up for a long stretch of underuse...

lindsayad replied:
Yea Im glad I got to see two gripping hands today. We have made forward declaration headers in quite a few places throughout our software stack to try and help with changing class template signatures, etc. But anyway I think what's here is fine. Marking this as resolved

GiudGiud replied:
imo we could chose to avoid tensor_value.h and vector_value.h as the typedefs are simple.
But for the ranges the typedefs are complicated and I think we even need to include other heads to be able to define them.

GiudGiud committed:
Sort out the always-included libmesh headers by reason for inclusion
Try to remove mesh_tools.h for now

GiudGiud commented:
Rebased on next to add newsletter entry. Coverage failure may happen but should be fine since this PR does not add new source, only some prefixes and switching around includes
